tests/testthat/test-indentation_linter.R

```diff
@@ -0,0 +1,423 @@
+  linter <- indentation_linter(indent = 2L)
+
+  expect_lint(
+    trim_some("
+      lapply(1:10, function(i) {
+        i %% 2
+      })
+    "),
+    NULL,
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      lapply(1:10, function(i) {
+       i %% 2  # indentation is only 1 character
+      })
+    "),
+    "Indentation",
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      lapply(1:10, function(i) {
+          i %% 2
+      })
+    "),
+    NULL,
+    indentation_linter(indent = 4L)
+  )
+
+  expect_lint(
+    trim_some("
+      lapply(1:10, function(i) {
+        i %% 2  # indentation is only 2 characters
+      })
+    "),
+    "Indentation",
+    indentation_linter(indent = 4L)
+  )
+
+  # ugly code, but still correctly indented
+  expect_lint(
+    trim_some("
+      list(
+           1,
+           2)
+    "),
+    NULL,
+    linter
+  )
+
+  # comments do not trigger hanging indent rule
+  expect_lint(
+    trim_some("
+      list( # comment
+        ok
+      )
+    "),
+    NULL,
+    linter
+  )
+
+  # assignment triggers indent
+  expect_lint(
+    trim_some("
+      a <-
+        expr(
+          42
+        )
+    "),
+    NULL,
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      if (cond)
+        code
+
+      if (cond) code else code2
+
+      if (cond) {
+        code
+      } else
+        code
+
+      if (cond) {
+        code
+      } else {
+        code
+      }
+    "),
+    NULL,
+    linter
+  )
+})
+
+test_that("indentation linter flags improper closing curly braces", {
+  linter <- indentation_linter(indent = 2L)
+  expect_lint(
+    trim_some("
+      lapply(1:10, function(i) {
+        {
+          i %% 2
+        }
+      })
+    "),
+    NULL,
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      lapply(1:10, function(i) {
+        i %% 2
+        } # closing curly doesn't return to parent indentation
+      )
+    "),
+    "Indentation",
+    linter
+  )
+})
+
+test_that("function argument indentation works in tidyverse-style", {
+  linter <- indentation_linter()
+  expect_lint(
+    trim_some("
+      function(a = 1L,
+               b = 2L) {
+        a + b
+      }
+    "),
+    NULL,
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      test <- function(a = 1L,
+                       b = 2L) {
+        a + b
+      }
+    "),
+    NULL,
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function(a = 1L,
+         b = 2L) {
+        a + b
+      }
+    "),
+    "Hanging",
+    linter
+  )
+
+  # This is a case for brace_linter
+  expect_lint(
+    trim_some("
+      function(a = 1L,
+               b = 2L)
+      {
+        a + b
+      }
+    "),
+    NULL,
+    linter
+  )
+})
+
+test_that("indentation with operators works", {
+  linter <- indentation_linter()
+  expect_lint(
+    trim_some("
+      a %>%
+        b()
+    "),
+    NULL,
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      (a + b + c) /
```

Comment:
i think there's some ambiguity on whether something like

```r
y ~
  x1 + 
    x2
```

should lint... on the one hand, it's like `~` and `+` independently induce an indent. on the other, all operators introduce one max indent.

FWIW, {styler} aligns with the choice here:

```r
styler::style_text("
y ~
  x1 +
    x2
")

# y ~
#   x1 +
#   x2
```

but maybe that's an option to explore? (probably best for follow-up)

Reply:
This seems like a peculiarity in the way `~` is parsed. Good catch, but should be a follow up I think.
See
``` r
lintr::lint(text = "y ~\n  x1 +\n  x2", linter = lintr::indentation_linter())
#> <text>:3:2: style: [indentation_linter] Indentation should be 4 spaces but is 2 spaces.
#>   x2
#>  ^~~
lintr::lint(text = "y +\n  x1 +\n  x2", linter = lintr::indentation_linter())
```

<sup>Created on 2022-10-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Filed #1718